repo/fsrepo/datastores.go

```diff
@@ -0,0 +1,316 @@
+	flatfs "gx/ipfs/QmUTshC2PP4ZDqkrFfDU4JGJFMWjYnunxPgkQ6ZCA2hGqh/go-ds-flatfs"
+	ds "gx/ipfs/QmVSase1JP7cq9QkPT46oNwdp9pT6kBkG3oqS14y3QcZjG/go-datastore"
+	mount "gx/ipfs/QmVSase1JP7cq9QkPT46oNwdp9pT6kBkG3oqS14y3QcZjG/go-datastore/syncmount"
+	ldbopts "gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb/opt"
+)
+
+// ConfigFromMap creates a new datastore config from a map
+type ConfigFromMap func(map[string]interface{}) (DatastoreConfig, error)
+
+type DiskSpec map[string]interface{}
+
+type DatastoreConfig interface {
+	// DiskSpec returns a minimal configuration of the datastore
+	// represting what is stored on disk.  Run time values are
+	// excluded.
+	DiskSpec() DiskSpec
+
+	// Create instantiate a new datastore from this config
+	Create(path string) (repo.Datastore, error)
+}
+
+func (spec DiskSpec) Bytes() []byte {
+	b, err := json.Marshal(spec)
+	if err != nil {
+		// should not happen
+		panic(err)
+	}
+	return bytes.TrimSpace(b)
+}
+
+func (spec DiskSpec) String() string {
+	return string(spec.Bytes())
+}
+
+var datastores map[string]ConfigFromMap
+
+func init() {
+	datastores = map[string]ConfigFromMap{
+		"mount":   MountDatastoreConfig,
+		"flatfs":  FlatfsDatastoreConfig,
+		"levelds": LeveldsDatastoreConfig,
+		"mem":     MemDatastoreConfig,
+		"log":     LogDatastoreConfig,
+		"measure": MeasureDatastoreConfig,
+	}
+}
+
+func AnyDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	which, ok := params["type"].(string)
+	if !ok {
+		return nil, fmt.Errorf("'type' field missing or not a string")
+	}
+	fun, ok := datastores[which]
+	if !ok {
+		return nil, fmt.Errorf("unknown datastore type: %s", which)
+	}
+	return fun(params)
+}
+
+type mountDatastoreConfig struct {
+	mounts []premount
+}
+
+type premount struct {
+	ds     DatastoreConfig
+	prefix ds.Key
+}
+
+func MountDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	var res mountDatastoreConfig
+	mounts, ok := params["mounts"].([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("'mounts' field is missing or not an array")
+	}
+	for _, iface := range mounts {
+		cfg, ok := iface.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("expected map for mountpoint")
+		}
+
+		child, err := AnyDatastoreConfig(cfg)
+		if err != nil {
+			return nil, err
+		}
+
+		prefix, found := cfg["mountpoint"]
+		if !found {
+			return nil, fmt.Errorf("no 'mountpoint' on mount")
+		}
+
+		res.mounts = append(res.mounts, premount{
+			ds:     child,
+			prefix: ds.NewKey(prefix.(string)),
+		})
+	}
+
+	return &res, nil
+}
+
+func (c *mountDatastoreConfig) DiskSpec() DiskSpec {
+	cfg := map[string]interface{}{"type": "mount"}
+	mounts := make([]interface{}, len(c.mounts))
+	for i, m := range c.mounts {
+		c := m.ds.DiskSpec()
+		if c == nil {
+			c = make(map[string]interface{})
+		}
+		c["mountpoint"] = m.prefix
+		mounts[i] = c
+	}
+	cfg["mounts"] = mounts
```

Comment:
This serializes the mount prefix as `"mountpoint":{"string":"/blocks"}`. Should just be the string.